### PR TITLE
[04] Pig Latin

### PR DESCRIPTION
**Pull request by lprogg**
## Description
Solved exercise 04.

## Related tasks
None.

## Steps to Test or Reproduce
None.

### CODE REVIEW

File: 04_pig_latin/pig_latin.rb [hunk 1]
```diff
@@ -0,0 +1,39 @@
+def logic(str)
+  vowels = %w[a e i o u]
+
+  while !vowels.include?(str[0])
```
Comment (alex-damian-negru) [inline 04_pig_latin/pig_latin.rb]: In the Ruby world we don't really use `loop`, `while`, and others, we rather use `each`, `map` and so on (basically looping functions that are applied directly on the objects).

Can you think of a solution that uses `each` or `map`?

File: 04_pig_latin/pig_latin.rb [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+    else
+      str = str[1..] << str[0]
+    end
+  end
+
+  str << 'ay'
+end
+
+def translate(str)
+  str.split(' ').map { |item| logic(item) }.join(' ')
```
Comment (alex-damian-negru) [inline 04_pig_latin/pig_latin.rb]: `split` uses a whitespace by default. https://apidock.com/ruby/String/split
```ruby
'abc abc'.split => ['abc', 'abc']
```

File: 04_pig_latin/pig_latin.rb [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+end
+
+def capitalize(str)
```
Comment (alex-damian-negru) [inline 04_pig_latin/pig_latin.rb:19]: Can you make this a one-line function? :upside_down_face:

File: 04_pig_latin/pig_latin_spec.rb [hunk 4]
```diff
@@ -68,5 +68,28 @@
   # Test-driving bonus:
   # * write a test asserting that capitalized words are still capitalized (but with a different initial capital letter, of course)
   # * retain the punctuation from the original phrase
+end
+
+describe "capitalize" do
```
Comment (alex-damian-negru) [inline 04_pig_latin/pig_latin_spec.rb:73]: Nice :100:. Good job writing new tests.

File: 04_pig_latin/pig_latin_spec.rb [hunk 5]
```diff
@@ -68,5 +68,28 @@
...
+  it "retains capitalized words" do
+    s = capitalize("Square")
```
Comment (alex-damian-negru) [inline 04_pig_latin/pig_latin_spec.rb]: What's `s`? Please use more descriptive variable names. 

Longer & clearer names are better than shorter & unclear ones.

**Review by alex-damian-negru — CHANGES_REQUESTED** [review]

File: 04_pig_latin/pig_latin.rb [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+end
+
+def translate(str)
+  str.split.map { |item| logic(item) }.join(' ')
+end
+
+def capitalize(str)
+  str.split.map { |item|
+    item == item.capitalize ? translate(item).capitalize : translate(item)
+  }.join(' ')
```
Comment (alex-damian-negru) [inline 04_pig_latin/pig_latin.rb]: Still ain't a one-liner, but that's close enough.

Note than the ruby style is to use `{ }` notation when the code is on the same line, e.g.
```ruby
'a b c'.split.map { |letter| letter.upcase }
```
while for multi-line we use `do ... end` notation:
```ruby
'a b c'.split.map do |letter|
  letter
   .upcase
   .downcase
   .whatever
end
```

**Review by alex-damian-negru — CHANGES_REQUESTED** [review]
Let's follow this convention for the PR namings:
```
[04] Pig Latin
```
Ensure to use the _entire_ name of the exercise, not a part of it.

File: 04_pig_latin/pig_latin.rb [hunk 7]
```diff
@@ -0,0 +1,35 @@
+def logic(str)
+  vowels = %w[a e i o u]
+
+  until vowels.include?(str[0])
```
Comment (alex-damian-negru) [inline 04_pig_latin/pig_latin.rb:4]: `until` is a better choice than `while/loop`

File: 04_pig_latin/pig_latin.rb (hunk 1 above)
Comment (lprogg) [reply 04_pig_latin/pig_latin.rb]: I used `until` construct.

**Pull request merged**